nervosnetwork/ckb
# feat(rpc): add a rpc to check if the ckb is ready

### Commits
- [chore: add a macro to construct `ExitCode=Failure` more efficient](2295/commits/71214f9)
  And replace few `expect` by `ExitCode=Failure`.
- [chore: add a dummy migration under feature "test-migration" for testing the delay by migrating](2295/commits/999bf5c)
- [feat(rpc): add a controller to support refresh configurations without restart](2295/commits/36b34b1)
  - So we can enable or disable RPC modules without stop CKB. (This feature will be added in other PRs.)
  - [TODO] Should be asynchronous, depend on global runtime.
- [feat(rpc): add a rpc to check if the ckb is ready](2295/commits/74d4936)
  - I used the simple RPC method `is_ready() -> bool` for keeping compatible as long as possible.
    Complex methods often are easy to break.
  - [**TODO**, Someday] A completed RPC to fetch all states of inner services should return:
    ```js
    {
      "result": {
        "tx-pool": "Ready",
        "net": "Booting",
        "store": "Migrating",
        "rpc": {
          "System": "Ready",
          "Debug": "Disabled",
          "Chain": "Booting",
          // ... ...
        }
      // ... ...
      }
    }
    ```
    Current architecture is not suitable to  implement this.

### Test

Since the test requires build CKB twice, so I didn't add this test into CI.

- Build a normal version CKB executable, or download an old version from GitHub releases.
- Run `ckb init --chain dev -C dev-chain` to initialize a develop chain.
- Run `ckb run -C dev-chain` to initialize data of the develop chain, then use `Ctrl + C` to stop the CKB.
- Set the feature "test-migration" in `ckb-shared` to be default, then re-build the CKB executable.
  A dummy migration for testing the delay by migrating was added, but it was controlled by the feature "test-migration".
- Run `ckb run -C dev-chain` to start the develop chain with new executable.
- Call RPC `is_ready`:
  - In first 30 seconds (the dummy migration will be blocked for 30 seconds), the state should be `is_ready=false`;
  - Then, the state should be `is_ready=true`;
